pkg/executor/sortexec/parallel_sort_spill_helper.go

```diff
@@ -0,0 +1,225 @@
+}
+
+func (p *parallelSortSpillHelper) isSpillNeeded() bool {
+	p.cond.L.Lock()
+	defer p.cond.L.Unlock()
+	return p.spillStatus == needSpill
+}
+
+func (p *parallelSortSpillHelper) isSpillTriggered() bool {
+	p.cond.L.Lock()
+	defer p.cond.L.Unlock()
+	return len(p.sortedRowsInDisk) > 0
+}
+
+func (p *parallelSortSpillHelper) setInSpilling() {
+	p.cond.L.Lock()
+	defer p.cond.L.Unlock()
+	p.spillStatus = inSpilling
+}
+
+func (p *parallelSortSpillHelper) setNeedSpillNoLock() {
+	p.spillStatus = needSpill
+}
+
+func (p *parallelSortSpillHelper) setNotSpilled() {
+	p.cond.L.Lock()
+	defer p.cond.L.Unlock()
+	p.spillStatus = notSpilled
+}
+
+func (p *parallelSortSpillHelper) spill() (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			err = util.GetRecoverError(r)
+		}
+	}()
+
+	select {
+	case <-p.finishCh:
+		return nil
+	default:
+	}
+
+	workerNum := len(p.sortExec.Parallel.workers)
+	workerWaiter := &sync.WaitGroup{}
+	workerWaiter.Add(workerNum)
+	sortedRowsIters := make([]*chunk.Iterator4Slice, workerNum)
+	for i := 0; i < workerNum; i++ {
+		go func(idx int) {
+			defer func() {
+				if r := recover(); r != nil {
+					processPanicAndLog(p.errOutputChan, r)
+				}
+				workerWaiter.Done()
+			}()
+
+			sortedRowsIters[idx] = chunk.NewIterator4Slice(nil)
+			sortedRowsIters[idx].Reset(p.sortExec.Parallel.workers[idx].sortLocalRows())
+			injectParallelSortRandomFail(200)
+		}(i)
+	}
+
+	workerWaiter.Wait()
+	p.setInSpilling()
+
+	// Spill is done, broadcast to wake up all sleep goroutines
+	defer p.cond.Broadcast()
+	defer p.setNotSpilled()
+
+	totalRows := 0
+	for i := range sortedRowsIters {
+		totalRows += sortedRowsIters[i].Len()
+	}
+
+	merger := newMultiWayMerger(sortedRowsIters, p.lessRowFunc)
+	merger.init()
+	return p.spillImpl(merger)
+}
+
+func (p *parallelSortSpillHelper) releaseMemory() {
+	totalReleasedMemory := int64(0)
+	for _, worker := range p.sortExec.Parallel.workers {
+		totalReleasedMemory += worker.totalMemoryUsage
+		worker.totalMemoryUsage = 0
+	}
+	p.sortExec.memTracker.Consume(-totalReleasedMemory)
+}
+
+func (p *parallelSortSpillHelper) spillTmpSpillChunk(inDisk *chunk.DataInDiskByChunks) error {
+	err := inDisk.Add(p.tmpSpillChunk)
+	if err != nil {
+		return err
+	}
+	p.tmpSpillChunk.Reset()
+	return nil
+}
+
+func (p *parallelSortSpillHelper) spillImpl(merger *multiWayMerger) error {
+	logutil.BgLogger().Info(spillInfo, zap.Int64("consumed", p.bytesConsumed.Load()), zap.Int64("quota", p.bytesLimit.Load()))
+	p.tmpSpillChunk.Reset()
+	inDisk := chunk.NewDataInDiskByChunks(p.fieldTypes)
+	inDisk.GetDiskTracker().AttachTo(p.sortExec.diskTracker)
+
+	spilledRowChannel := make(chan chunk.Row, 10000)
+	go func() {
+		defer func() {
+			if r := recover(); r != nil {
+				processPanicAndLog(p.errOutputChan, r)
+			}
+			close(spilledRowChannel)
+		}()
+
+		injectParallelSortRandomFail(200)
+
+		for {
+			row := merger.next()
+			if row.IsEmpty() {
+				break
+			}
+			spilledRowChannel <- row
+		}
+	}()
```

Comment:
Is executing concurrently here necessary?

Reply:
> Is executing concurrently here necessary?

Yes, as merge and disk operation could be concurrently executed.